osf/external/spam/tasks.py

```diff
@@ -33,18 +35,42 @@ def reclassify_domain_references(notable_domain_id, current_note, previous_note)
 def check_resource_for_domains(guid, content):
     from osf.models import Guid, NotableDomain, DomainReference
     resource = Guid.load(guid).referent
-    domains = {match.group('domain') for match in re.finditer(DOMAIN_REGEX, content)}
     spammy_domains = []
     referrer_content_type = ContentType.objects.get_for_model(resource)
-    for domain in domains:
-        domain, _ = NotableDomain.objects.get_or_create(domain=domain)
-        if domain.note == NotableDomain.Note.EXCLUDE_FROM_ACCOUNT_CREATION_AND_CONTENT:
-            spammy_domains.append(domain.domain)
+    for domain in _extract_domains(content):
+        notable_domain, _ = NotableDomain.objects.get_or_create(domain=domain)
+        if notable_domain.note == NotableDomain.Note.EXCLUDE_FROM_ACCOUNT_CREATION_AND_CONTENT:
+            spammy_domains.append(notable_domain.domain)
         DomainReference.objects.get_or_create(
-            domain=domain,
+            domain=notable_domain,
             referrer_object_id=resource.id,
             referrer_content_type=referrer_content_type,
-            defaults={'is_triaged': domain.note != NotableDomain.Note.UNKNOWN}
+            defaults={'is_triaged': notable_domain.note != NotableDomain.Note.UNKNOWN}
         )
     if spammy_domains:
         resource.confirm_spam(save=True, domains=list(spammy_domains))
+
+def _extract_domains(content):
+    extracted_domains = set()
+    for match in DOMAIN_REGEX.finditer(content):
+        domain = match.group('domain')
+        if not domain or domain in extracted_domains:
+            continue
+
+        protocol = match.group('protocol') or 'https://'
+        www = match.group('www') or ''
+        constructed_url = f'{protocol}{www}{domain}'
+
+        try:
+            response = requests.head(constructed_url)
+        except requests.exceptions.ConnectionError:
+            continue
+
+        # Follow 302 in case of link shorteners
+        if response.status_code == 302:
+            domain = DOMAIN_REGEX.match(response.headers['location']).group('domain')
```

Comment:
Is a response header guaranteed to have a `location`? What if a spammer deliberately puts domain A and B in a project, and A would give a 302 response code with no `location` in the header? Would it break domain extraction?

Reply:
Good catch. Did some research and it looks like:
A) Should be checking for a larger variety of codes (301, 302, 303, 307, 308)
B) The `location` attribute is specified as something the server SHOULD generate, but it's impossible to enforce it. Most browsers would probably get terribly confused if it isn't there, but we'll handle the case anyway.

Reply:
I'm thinking as a spammer who reads this code. And I'd spin up a server to host domain A and it returns 301/302/303 etc. but doesn't included `location` in the response header. Then the tasks exits with a `KeyError` thrown. This is the worst part about having spam-related code in an open-sourced repo unfortunately.